### PR TITLE
tdnf: init at 3.5.6

### PR DESCRIPTION
**Pull request by malt3**
## Description of changes

Add tdnf, a dnf variant similar to microdnf.
This package manager is used to assemble the linux distributions VMWare Photon OS and Microsoft's cbl-mariner / Azure Linux.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/td/tdnf/package.nix [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    "-DCMAKE_INSTALL_SYSCONFDIR=$out/etc"
+    "-DSYSTEMD_DIR=$out/lib/systemd/system"
+  ];
+
+  # error: format not a string literal and no format arguments [-Werror=format-security]
+  hardeningDisable = [ "format" ];
+
+  postPatch = ''
+    substituteInPlace CMakeLists.txt \
+      --replace 'SYSCONFDIR /etc' 'SYSCONFDIR $out/etc' \
```
Comment (msanft) [inline pkgs/by-name/td/tdnf/package.nix:56]: Any reason for not using `--replace-fail` here? Seems brittle to me..

File: pkgs/by-name/td/tdnf/package.nix [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+      --replace '/etc/motdgen.d' '$out/etc/motdgen.d'
+    substituteInPlace client/tdnf.pc.in \
+      --replace-fail 'libdir=''${prefix}/@CMAKE_INSTALL_LIBDIR@' 'libdir=@CMAKE_INSTALL_FULL_LIBDIR@'
+    substituteInPlace tools/cli/lib/tdnf-cli-libs.pc.in \
+      --replace-fail 'libdir=''${prefix}/@CMAKE_INSTALL_LIBDIR@' 'libdir=@CMAKE_INSTALL_FULL_LIBDIR@'
+  '';
+
+  meta = with lib; {
```
Comment (msanft) [inline pkgs/by-name/td/tdnf/package.nix:64]: Using `with lib;` is an anti-pattern.

File: pkgs/by-name/td/tdnf/package.nix [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    description = "Tiny Dandified Yum";
+    homepage = "https://github.com/vmware/tdnf";
+    changelog = "https://github.com/vmware/tdnf/releases/tag/v${finalAttrs.version}";
+    license = [ licenses.gpl2 licenses.lgpl21 ];
```
Comment (msanft) [inline pkgs/by-name/td/tdnf/package.nix:68]:
```suggestion
    license = with lib.licenses; [ gpl2 lgpl21 ];
```

File: pkgs/by-name/td/tdnf/package.nix [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    maintainers = with maintainers; [ malt3 ];
```
Comment (msanft) [inline pkgs/by-name/td/tdnf/package.nix:69]:
```suggestion
    maintainers = [ lib.maintainers.malt3 ];
```

File: pkgs/by-name/td/tdnf/package.nix [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    mainProgram = "tdnf";
+    platforms = platforms.unix;
```
Comment (msanft) [inline pkgs/by-name/td/tdnf/package.nix:71]:
```suggestion
    platforms = lib.platforms.unix;
```

**Review by msanft — CHANGES_REQUESTED** [review]

**Comment by msanft** [comment]
Result of `nixpkgs-review pr 307698` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>tdnf</li>
  </ul>
</details>

**Comment by msanft** [comment]
Package seems to function properly

**Review by msanft — APPROVED** [review]
Thank you!

**Review by katexochen — COMMENTED** [review]
I think we should remove `bin/jsondumptest`, seems to be a test output?

**Review by katexochen — APPROVED** [review]

**Pull request merged**